test/metrics/query.go

```diff
@@ -0,0 +1,260 @@
+package metrics
+
+// package metrics provides the following exported functions:
+//     LocatePrometheus
+//     CreateAnnotatedCABundleInjectionConfigMap
+//     PodExecCreate
+//     RunQueryFromPod
+// These functions can validate prometheus metrics
+// and also will validate the current cluster CA
+// issued through service-ca-operator.
+// To only validate the CA, one can pass the watchdog alert to RunQueryFromPod,
+// as this alert should always be firing in a healthy cluster:
+//     `ALERTS{alertstate="firing",alertname="Watchdog",severity="none"} == 1`
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"os/exec"
+	"strings"
+	"syscall"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	uexec "k8s.io/utils/exec"
+
+	"github.com/prometheus/common/model"
+)
+
+const (
+	prometheusBaseURL         = "https://prometheus-k8s.openshift-monitoring.svc:9091"
+	serviceCAInjectAnnotation = "service.beta.openshift.io/inject-cabundle"
+)
+
+func LocatePrometheus(client *kubernetes.Clientset) (string, error) {
+	_, err := client.CoreV1().Services("openshift-monitoring").Get("prometheus-k8s", metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	var bearerToken string
+	secrets, err := client.CoreV1().Secrets("openshift-monitoring").List(metav1.ListOptions{})
+	if err != nil {
+		return "", fmt.Errorf("could not list secrets in openshift-monitoring namespace")
+	}
+	for _, secret := range secrets.Items {
+		if secret.Type != corev1.SecretTypeServiceAccountToken {
+			continue
+		}
+		if !strings.HasPrefix(secret.Name, "prometheus-") {
+			continue
+		}
+		bearerToken = string(secret.Data[corev1.ServiceAccountTokenKey])
+		break
+	}
+	if len(bearerToken) == 0 {
+		return "", fmt.Errorf("error getting bearer token from prometheus service account")
+	}
+	return bearerToken, nil
+}
+
+type prometheusResponse struct {
+	Status string                 `json:"status"`
+	Data   prometheusResponseData `json:"data"`
+}
+
+type prometheusResponseData struct {
+	ResultType string       `json:"resultType"`
+	Result     model.Vector `json:"result"`
+}
+
+// CreateAnnotatedCABundleConfigMap results in configmap injected with CA bundle issued by service-ca-operator
+func CreateAnnotatedCABundleInjectionConfigMap(client *kubernetes.Clientset, configMapName, nsName string) error {
```

Comment:
Worth noting down that the method also waits for that CM to be created.

Reply:
`CreateAndWaitFor...` would be a better name

Reply:
updated